### Examples

In [29]:
%display latex

#### Small nontrivial algebraic example

In [30]:
A.<x> = QQ[]
a = 3*x-4
b = 2*(x^2-3*x+2) # a/b has rational residues
bb = b/2 # 2 is the content of b

In [31]:
%time sigma(a,b)

CPU times: user 2.25 ms, sys: 24 μs, total: 2.28 ms
Wall time: 2.29 ms


3269291615

In [32]:
%time sigma(a,bb)

CPU times: user 2.02 ms, sys: 0 ns, total: 2.02 ms
Wall time: 2.02 ms


265

Highlights the importance of dividing by content, running Honda(a,b) will never finish, but Honda(a,bb) will and with the same output.

In [33]:
%time Honda(a,bb)

CPU times: user 27.9 ms, sys: 9 μs, total: 27.9 ms
Wall time: 27.3 ms


('algebraic', 0)

In [34]:
%time CompFactor(a,b)

CPU times: user 954 μs, sys: 0 ns, total: 954 μs
Wall time: 959 μs


'algebraic'

For algebraic examples, factorization is faster even for small examples.

#### $1/(x^2-c)$

In [36]:
A.<x> = QQ[]

In [37]:
a = 1
b = x^2 - 3818929

This is the smallest quadratic example for which all $p$-curvatures will vanish (and $2$ is exceptional) for $p<47$, and the $47$ will not vanish.

In [38]:
%time Honda(a,b)

CPU times: user 7.67 ms, sys: 0 ns, total: 7.67 ms
Wall time: 7.32 ms


('transcendental', 47)

In [39]:
a = 2*x+1
b = x^2+x+1

In [40]:
sigma(a,b)

3851774

In [41]:
%time Timepcurvs(a,b,10^5)

CPU times: user 30.3 s, sys: 14.8 ms, total: 30.3 s
Wall time: 30.3 s


Computing all $p$-curvatures up to $\sigma = 3851774$ is not a possibility.

#### Small transcendental example

In [54]:
%display plain

In [55]:
a = 7*x^2-3*x-4
b = 2*(x^3+2*x^2-3*x+2)

In [56]:
sigma(a,b)

6393533718461168896735722504

In [57]:
%time Honda(a,b)

CPU times: user 5.07 ms, sys: 1 μs, total: 5.07 ms
Wall time: 4.96 ms


('transcendental', 5)

In [58]:
%time CompFactor(a,b)

CPU times: user 484 μs, sys: 2 μs, total: 486 μs
Wall time: 488 μs


'transcendental'

#### Large random (hence transcendental) examples

In [59]:
import time

In [63]:
def RandComp(n,H,S):
    """
    n, H, S are positive integers.
    Creates a sample of S rational functions a/b with integer coefficients between -H and H, deg(b)=n, deg(a)<n.
    For each of them computes Honda(a,b) and CompFactor(a,b), and measures the time.
    Returns the average runtime of each function, and the maximal prime that allowed Honda to conclude on transcendence.
    """
    A.<x> = QQ[]
    THond = 0
    Tfact = 0
    max_p = 1
    for i in range(S):
        a = A([randint(-H,H) for i in range(n)])
        b = A([randint(-H,H) for i in range(n)]+[randint(1,H)])
        t0 = time.time()
        s, p = Honda(a,b)
        dt = time.time()-t0
        THond += dt
        if p>max_p:
            max_p=p
        t1 = time.time()
        CompFactor(a,b)
        dt = time.time()-t1
        Tfact += dt
    return THond/S, max_p, Tfact/S

In [65]:
n=20
H=2^20
S=50

In [66]:
%time RandComp(n,H,S)

CPU times: user 446 ms, sys: 1.01 ms, total: 447 ms
Wall time: 446 ms


(0.0020928525924682616, 13, 0.006773462295532226)

The function RandComp was used to make the table of comparisons on large example.

#### Galactic examples

We are looking for the limits of our algorithm (that is heights and degrees for which computing only a few $p$-curvatures takes more than a few seconds).
The approach is the same as in the previous block of cells, random hence transcendental examples, but this time on samples of very small size.

In [73]:
H = 2^(128)
n = 1000

In [74]:
a = A([randint(-H,H) for i in range(n)])
b = A([randint(-H,H) for i in range(n)]+[randint(1,H)])

In [75]:
%time Honda(a,b)

CPU times: user 10.7 s, sys: 30 ms, total: 10.7 s
Wall time: 10.7 s


('transcendental', 3)